File: leetcode/word-search.java
//https://leetcode.com/problems/word-search/

class Solution {
    public boolean exist(char[][] board, String word) {
        boolean[][] visited = new boolean[board.length][board[0].length];
        
        for(int i = 0; i<board.length; i++){
            for(int j = 0; j<board[0].length; j++){
                if(word.charAt(0) == board[i][j]){
                    if(search(board, word, i,j,0, visited)) return true;
                }
            }
        }
        return false;
    }
    
    private boolean search(char[][] board, String word, int row, int col,int index, boolean[][] visited){
        if(index == word.length()) return true;
        if(row < 0 || row> board.length-1 || col<0 || col>board[0].length-1 || board[row][col]!=word.charAt(index) || visited[row][col]) return false;
        
        visited[row][col] = true;
        boolean continueSearch = 
            search(board, word, row+1, col, index+1, visited) ||
            search(board, word, row-1, col, index+1, visited) ||
            search(board, word, row, col+1, index+1, visited) ||
            search(board, word, row, col-1, index+1, visited);

        visited[row][col] = false;
        return continueSearch;
    }
}

File: leetcode/word_search.py
# https://leetcode.com/problems/word-search/description/


class Solution:
    def exist(self, board: list[list[str]], word: str) -> bool:
        if len(word) > len(board) * len(board[0]):
            return False

        for r in range(len(board)):
            for c in range(len(board[0])):
                if board[r][c] == word[0]:
                    if self.search(board, r, c, 1, word):
                        return True
        return False

    def search(self, board: list[list[str]], row: int, col: int, index: int, word: str) -> bool:
        if index == len(word):
            return True

        board[row][col] = "-"
        found: bool = False
        # search above
        if row > 0 and board[row - 1][col] == word[index]:
            found = self.search(board, row - 1, col, index + 1, word)

        if not found:
            # search right
            if col < len(board[0]) - 1 and board[row][col + 1] == word[index]:
                found = self.search(board, row, col + 1, index + 1, word)
            if not found:
                # search below
                if row < len(board) - 1 and board[row + 1][col] == word[index]:
                    found = self.search(board, row + 1, col, index + 1, word)
                # search left
                if not found and col > 0 and board[row][col - 1] == word[index]:
                    found = self.search(board, row, col - 1, index + 1, word)

        board[row][col] = word[index - 1]
        return found
